File: snippets/fsharp/System/Type/IsAssignableFrom/isassignablefrom_ex1.fs
module isassignablefrom_ex1

// <Snippet1>
open System
open System.Reflection
open System.Reflection.Emit

type A() = class end

let domain = AppDomain.CurrentDomain
let assemName = AssemblyName()
assemName.Name <- "TempAssembly"

// Define a dynamic assembly in the current application domain.
let assemBuilder = 
    domain.DefineDynamicAssembly(assemName, AssemblyBuilderAccess.Run)

// Define a dynamic module in this assembly.
let moduleBuilder = 
    assemBuilder.DefineDynamicModule "TempModule"

let b1 = moduleBuilder.DefineType("B", TypeAttributes.Public, typeof<A>)
printfn $"{typeof<A>.IsAssignableFrom b1}"
// The example displays the following output:
//        True
// </Snippet1>

File: snippets/csharp/System/Type/IsAssignableFrom/isassignablefrom_ex1.cs
// <Snippet1>
using System;
using System.Reflection;
using System.Reflection.Emit;

public class A
{ }

public class IsAssignableFromExample1
{
    public static void Run()
    {
        AssemblyName assemName = new()
        {
            Name = "TempAssembly"
        };

        // Define a dynamic assembly.
        AssemblyBuilder assemBuilder = AssemblyBuilder.DefineDynamicAssembly(
            assemName, AssemblyBuilderAccess.Run);

        // Define a dynamic module in this assembly.
        ModuleBuilder moduleBuilder = assemBuilder.DefineDynamicModule("TempModule");

        TypeBuilder b1 = moduleBuilder.DefineType("B", TypeAttributes.Public, typeof(A));
        Console.WriteLine(typeof(A).IsAssignableFrom(b1));
    }
}
// The example displays the following output:
//        True
// </Snippet1>
